lib/linter.js

```diff
@@ -814,27 +814,35 @@ class Linter extends EventEmitter {
 
             // if this result is from a parseForESLint() method, normalize
             if (parseResult && parseResult.ast) {
-                ast = parseResult.ast;
-            } else {
-                ast = parseResult;
-                parseResult = null;
-            }
 
-            if (ast) {
-                this.sourceCode = new SourceCode(text, ast);
+                // Save all returned values of `parse()` because this source code object can be reused.
+                // This ensures the same result when reused.
+                this.sourceCode = new SourceCode(
+                    text,
+                    parseResult.ast,
+                    parseResult.services,
+                    parseResult.scope,
+                    parseResult.visitorKeys
+                );
+            } else if (parseResult) {
+                this.sourceCode = new SourceCode(
+                    text,
+                    parseResult
+                );
+            } else {
+                this.sourceCode = null;
             }
 
         } else {
             this.sourceCode = textOrSourceCode;
-            ast = this.sourceCode.ast;
         }
 
         // if espree failed to parse the file, there's no sense in setting up rules
-        if (ast) {
+        if (this.sourceCode) {
```

Comment:
Should this include a check for `this.sourceCode.ast` to match the previous behaviour?